### ISSUE TITLE
ScaledDotProductAttention not returning attention properly.

### ISSUE THREAD

**Issue by deepettas** [CyberZHG/keras-self-attention#37]
I have used your MultiheadAttention model on a complex input, outputting attention/self attention with success.
After that, I have tried experimenting with the ScaledDotProductAttention model that is used from Multihead with the aim to increase interpretability and decrease inference/training times.

While building a simple example such as the one bellow, I was unable to extract a correct attention vector even after training the model. 
**The attention vector was always:  [0.99, 0.99, 0.99, .... , 0.99] for single dimentional input.**

Any clues to what I am doing that prevents the model outputting a correct attention vector?

Reference code bellow:

```python
from __future__ import absolute_import, division, print_function, unicode_literals
import os

import tensorflow as tf

import cProfile
import keras
import keras.backend as K
from keras_self_attention import ScaledDotProductAttention
import tensorflow as tf
import numpy as np
import keras


input_query = keras.layers.Input(
    shape=(9,),
    name='Input-Q',
)
input_key = keras.layers.Input(
    shape=(1, ),
    name='Input-K',
)
input_value = keras.layers.Input(
    shape=(1, ),
    name='Input-V',
)
emb = keras.layers.Embedding(input_dim=20, output_dim=10)

query = emb(input_query)
key = emb(input_key)
value = emb(input_value)

att_layer, a = ScaledDotProductAttention(
    name='ScaledDotProductAttention',
    return_attention= True
)([query, key, value])

att_layer = keras.layers.GlobalAveragePooling1D(name='AttAvg')(att_layer)

query = keras.layers.GlobalAveragePooling1D()(query)

query = keras.layers.Reshape(target_shape=(1, 10))(query)

att_layer = keras.layers.Reshape(target_shape=(1, 10))(att_layer)

y = keras.layers.concatenate([att_layer, query])
y = keras.layers.Reshape(target_shape=(20,))(y)
y = keras.layers.Dense(300, activation='relu')(y)

y = keras.layers.Dense(1, activation='sigmoid', name='output_sigmoid')(y)


model = keras.models.Model(
    inputs=[input_query, input_key, input_value], outputs=y)
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy'],
)
model.summary()
```

    Attention is:  Tensor("ScaledDotProductAttention_2/truediv_1:0", shape=(?, 9, 1), dtype=float32) Tensor("ScaledDotProductAttention_2/MatMul_1:0", shape=(?, 9, 10), dtype=float32) Tensor("embedding_4_2/GatherV2:0", shape=(?, 1, 10), dtype=float32)
    __________________________________________________________________________________________________
    Layer (type)                    Output Shape         Param #     Connected to                     
    ==================================================================================================
    Input-Q (InputLayer)            (None, 9)            0                                            
    __________________________________________________________________________________________________
    Input-K (InputLayer)            (None, 1)            0                                            
    __________________________________________________________________________________________________
    Input-V (InputLayer)            (None, 1)            0                                            
    __________________________________________________________________________________________________
    embedding_4 (Embedding)         multiple             200         Input-Q[0][0]                    
                                                                     Input-K[0][0]                    
                                                                     Input-V[0][0]                    
    __________________________________________________________________________________________________
    ScaledDotProductAttention (Scal [(None, 9, 10), (Non 0           embedding_4[0][0]                
                                                                     embedding_4[1][0]                
                                                                     embedding_4[2][0]                
    __________________________________________________________________________________________________
    AttAvg (GlobalAveragePooling1D) (None, 10)           0           ScaledDotProductAttention[0][0]  
    __________________________________________________________________________________________________
    global_average_pooling1d_3 (Glo (None, 10)           0           embedding_4[0][0]                
    __________________________________________________________________________________________________
    reshape_7 (Reshape)             (None, 1, 10)        0           AttAvg[0][0]                     
    __________________________________________________________________________________________________
    reshape_6 (Reshape)             (None, 1, 10)        0           global_average_pooling1d_3[0][0] 
    __________________________________________________________________________________________________
    concatenate_3 (Concatenate)     (None, 1, 20)        0           reshape_7[0][0]                  
                                                                     reshape_6[0][0]                  
    __________________________________________________________________________________________________
    reshape_8 (Reshape)             (None, 20)           0           concatenate_3[0][0]              
    __________________________________________________________________________________________________
    dense_2 (Dense)                 (None, 300)          6300        reshape_8[0][0]                  
    __________________________________________________________________________________________________
    output_sigmoid (Dense)          (None, 1)            301         dense_2[0][0]                    
    ==================================================================================================
    Total params: 6,801
    Trainable params: 6,801
    Non-trainable params: 0
    __________________________________________________________________________________________________

# Data generation

```python
# Data Generation.

from random import randint
def get_q(m,n):
    
    q = []
    
    for i in range(n):
        a = []
        k = randint(1,10)
        for i in range(m):
            a.append(k)
        q.append(a)
    return q

def get_k(n):
    l = randint(1,10)
    k = []
    for i in range(n):
        k.append((l))
    
    return k



def potato_gun(iterations):
    i_q = np.array([get_q(1,9) for i in range(0,iterations)])
    i_k = np.array([get_q(1,1) for i in range(0,iterations)])


    i_y = []
    for q, v in zip(i_q, i_k):
        y = 0

        for row in q:
            if np.array_equal(v[0],row):
                y = 1
        i_y.append(y)
    
    return [np.reshape(i_q, newshape=(iterations, 9)), 
            np.reshape(i_k, newshape=(iterations,)), 
            np.reshape(i_k, newshape=(iterations,))], i_y


dummy_dataset = potato_gun(240000)
x_train, y_train = (dummy_dataset[0],dummy_dataset[1])
dummy_dataset = potato_gun(80000)
x_test, y_test = (dummy_dataset[0],dummy_dataset[1])

```

# Simple rule: 
## If input[1] exists in the input[0] , y = 1, else y = 0

```python
idx = 5
print(f'If {x_train[0][idx]} contains {x_train[1][idx]}, Then y = {y_train[idx]}')
```

    If [ 9  8  9  6  4  2  6  7 10] contains [7], Then y = 1

# Training

```python
model.fit(x=x_train,
      y=y_train,
      batch_size=32,
      epochs=10,
      verbose=1,
      validation_data=(x_test, y_test),
      shuffle=True,
      class_weight=None,
      sample_weight=None
          )
```

    Train on 240000 samples, validate on 80000 samples
    Epoch 1/10
    240000/240000 [==============================] - 11s 47us/step - loss: 0.4857 - acc: 0.7541 - val_loss: 0.3153 - val_acc: 0.8487
    Epoch 2/10
    240000/240000 [==============================] - 11s 45us/step - loss: 0.1943 - acc: 0.9161 - val_loss: 0.0770 - val_acc: 0.9788
    Epoch 3/10
    240000/240000 [==============================] - 11s 46us/step - loss: 0.0113 - acc: 0.9986 - val_loss: 3.9026e-04 - val_acc: 1.0000
    Epoch 4/10
    240000/240000 [==============================] - 11s 46us/step - loss: 2.3738e-04 - acc: 1.0000 - val_loss: 9.9840e-05 - val_acc: 1.0000
    Epoch 5/10
    240000/240000 [==============================] - 12s 50us/step - loss: 6.5733e-05 - acc: 1.0000 - val_loss: 2.9516e-05 - val_acc: 1.0000
    Epoch 6/10
    240000/240000 [==============================] - 13s 56us/step - loss: 3.2543e-05 - acc: 1.0000 - val_loss: 1.5050e-05 - val_acc: 1.0000
    Epoch 7/10
    240000/240000 [==============================] - 13s 56us/step - loss: 2.5813e-06 - acc: 1.0000 - val_loss: 1.8419e-06 - val_acc: 1.0000
    Epoch 8/10
    240000/240000 [==============================] - 12s 50us/step - loss: 5.1268e-05 - acc: 1.0000 - val_loss: 2.1687e-06 - val_acc: 1.0000
    Epoch 9/10
    240000/240000 [==============================] - 12s 49us/step - loss: 5.6301e-07 - acc: 1.0000 - val_loss: 3.2817e-07 - val_acc: 1.0000
    Epoch 10/10
    240000/240000 [==============================] - 12s 51us/step - loss: 7.6280e-05 - acc: 1.0000 - val_loss: 9.8148e-07 - val_acc: 1.0000

    <keras.callbacks.History at 0x7fa5630a7630>

# Extracting the submodel & confirming the output shapes:

```python
target_layer = model.get_layer('ScaledDotProductAttention')
layer_output = target_layer.output
from keras.models import Model

submodel_emb = Model(inputs=model.input, outputs=layer_output)
print(submodel_emb.summary())
print(f'Shapes: \n Values: {submodel_emb.output_shape[0]}, Attention: {submodel_emb.output_shape[1]}')
```

    __________________________________________________________________________________________________
    Layer (type)                    Output Shape         Param #     Connected to                     
    ==================================================================================================
    Input-Q (InputLayer)            (None, 9)            0                                            
    __________________________________________________________________________________________________
    Input-K (InputLayer)            (None, 1)            0                                            
    __________________________________________________________________________________________________
    Input-V (InputLayer)            (None, 1)            0                                            
    __________________________________________________________________________________________________
    embedding_3 (Embedding)         multiple             200         Input-Q[0][0]                    
                                                                     Input-K[0][0]                    
                                                                     Input-V[0][0]                    
    __________________________________________________________________________________________________
    ScaledDotProductAttention (Scal [(None, 9, 10), (Non 0           embedding_3[0][0]                
                                                                     embedding_3[1][0]                
                                                                     embedding_3[2][0]                
    ==================================================================================================
    Total params: 200
    Trainable params: 200
    Non-trainable params: 0
    __________________________________________________________________________________________________
    None
    Shapes: 
     Values: (None, 9, 10), Attention: (None, 9, 1)

# Expected behaviour: 
### Model should export an attention vector containing attention weights summing up to 1 during .predict() 

# Observed behaviour:
# With input array [2,2,2,2,2,2,2,2,1] the attention should be directed in the last position of the input.

## Contrary to the expectations, the attention vector is [0.99, 0.99, 0.99, .... , 0.99]

```python
idx = 5
values, attention = submodel_emb.predict(
    (
        [np.array([[2]*8 + [1]]),
        np.array([1,]),
        np.array([1])]
    )
)
print(attention)
```

    [[[0.9999999]
      [0.9999999]
      [0.9999999]
      [0.9999999]
      [0.9999999]
      [0.9999999]
      [0.9999999]
      [0.9999999]
      [0.9999999]]]
